python/ql/src/semmle/python/templates/DjangoTemplate.qll

```diff
@@ -0,0 +1,49 @@
+/** Provides classes which model the `DjangoTemplate` package. */
+
+import python
+import semmle.python.web.HttpRequest
+import semmle.python.templates.Templates
+
+ClassValue theDjangoTemplateClass() { result = Value::named("django.template.Template") }
+
+/**
+ * Sink representng `django.template.Template` class instantiation argument.
+ *
+ *  from django.template import Template
+ *  template = Template(`sink`)
+ */
+class DjangoTemplateTemplateSink extends SSTISink {
+  override string toString() { result = "argument to Django.template()" }
+
+  DjangoTemplateTemplateSink() {
+    exists(CallNode call |
+      call.getFunction().pointsTo(theDjangoTemplateClass()) and
+      call.getArg(0) = this
+    )
+  }
+
+  override predicate sinks(TaintKind kind) { kind instanceof ExternalStringKind }
+}
+
+/**
+ *Sinks representng the django.template.Template class instantiation.
+ *
+ *  from django.template import engines
+ *
+ *  django_engine = engines["django"]
+ *  template = django_engine.from_string(`sink`)
+ */
+class DjangoTemplateEngineSink extends SSTISink {
+  override string toString() { result = "argument to Django.template()" }
+
+  // HELP : This does resolve `from_string`
```

Comment:
@RasmusWL @tausbn Can you please help me figure out why codeql fails to detect the `from_string` call here?

Reply:
I'm not sure if the code changed since you posted your comment, but `Value::named("")` doesn't make much sense. It tries to return the `Value` corresponding to the built-in with the name `""`, but there is no such built-in.

Reply:
The value was `django.template.Engine.from_string`. I think I also get why codeql failed to recognize that. Since the `django.template.Engine`module is not directly imported, the indirect reference is not recognized by codeql. Can you please confirm this?

Reply:
I think the issue here is that `django.template.Engine` is not a module but a class. The `Value::named(...)` interface is rather limited, in that it always assumes everything before the last `.` is the name of the module, and so `django.template.Engine.from_string` tries to match the `from_string` attribute in the `django.template.Engine` module (which doesn't exist).
Instead, what you probably want to do is pull out the `Engine` class separately as a `ClassValue`, and then take the `from_string` attribute on _that_, using the `attr` method.